core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -134,6 +134,8 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
   private val KAFKA_CLUSTER_ID: String = "kafka.cluster.id"
   private val KAFKA_BROKER_ID: String = "kafka.broker.id"
 
+  private val actionQueue = new ActionQueue
```

Comment:
Another approach would be for this queue to be per request thread instead of per server. That would simplify concurrency handling.

Reply:
The thread has to pass queue to method of ReplicManager/GroupCoordinator if queue is kept by thread. A lot of methods are included so that is a big patch. I prefer to keep small patch though it gets bigger now :(

Reply:
Note that the action queue is not only called by requests threads, but also by the expiration thread for certain delayed operations.

Reply:
Good point @junrao.